server/readonly/readonly_service.ts

```diff
@@ -0,0 +1,119 @@
+/*
+ *   Copyright OpenSearch Contributors
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+import { URL } from 'url';
+import {
+  Logger,
+  OpenSearchDashboardsRequest,
+  SessionStorageFactory,
+} from '../../../../src/core/server';
+import {
+  globalTenantName,
+  isPrivateTenant,
+  LOGIN_PAGE_URI,
+  CUSTOM_ERROR_PAGE_URI,
+} from '../../common';
+import { SecurityClient } from '../backend/opensearch_security_client';
+import { IAuthenticationType, OpenSearchAuthInfo } from '../auth/types/authentication_type';
+import { SecuritySessionCookie } from '../session/security_cookie';
+import { SecurityPluginConfigType } from '../index';
+import { ReadonlyService as BaseReadonlyService } from '../../../../src/core/server/security/readonly_service';
+
+export class ReadonlyService extends BaseReadonlyService {
+  protected static readonly ROUTES_TO_IGNORE: string[] = [LOGIN_PAGE_URI, CUSTOM_ERROR_PAGE_URI];
+
+  private readonly logger: Logger;
+  private readonly securityClient: SecurityClient;
+  private readonly auth: IAuthenticationType;
+  private readonly securitySessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>;
+  private readonly config: SecurityPluginConfigType;
+
+  constructor(
+    logger: Logger,
+    securityClient: SecurityClient,
+    auth: IAuthenticationType,
+    securitySessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>,
+    config: SecurityPluginConfigType
+  ) {
+    super();
+    this.logger = logger;
+    this.securityClient = securityClient;
+    this.auth = auth;
+    this.securitySessionStorageFactory = securitySessionStorageFactory;
+    this.config = config;
+  }
+
+  isAnonymousPage(request: OpenSearchDashboardsRequest) {
```

Comment:
Is it possible that the readonly service can be applied after a request is authenticated? If can be applied after auth is it possible that this method could not be needed? If this method is required then I'd like to see more test cases around this particular method because its not entirely clear what it means to be an anonymous page (anonymous page === route that doesn't require auth like the login screen?)

If the order of operations is:

1. Request is received
2. auth // if endpoint requires authentication
3. isReadOnly

then it would be guaranteed that `isReadOnly` is only called on authenticated requests.

It looks like `isReadOnly` is passthrough for the routes that do not require authentication so if it could be skipped then this logic may not be necessary.

Reply:
I figured someone might want to use the 'isReadonly' method for some custom logic. But it's kinda redundant unless you're planning to use it only within the 'capabilities.registerSwitcher' method.

In simple terms, you won't need it for what it's doing now. But, if you decide to go rogue and use it outside 'capabilities.registerSwitcher', it would be less confusing for a potential use case scenario. 

In summary:

> Is it possible that the readonly service can be applied after a request is authenticated

- for capabilities no
- for custom logic yes

> If it can be applied after auth

not for capabilities

>  its not entirely clear what it means to be an anonymous page (anonymous page === route that doesn't require auth like the login screen?)

exactly what you said - an anonymous page is a page that doesn't require an auth or auth is `optional` (like capabilities route 🫠)

It really depends if we would like to keep the possibility to call it outside `capabilities.registerSwitcher`. If the answer is yes then this method should stay, if not - we can entirely remove it.